ambry-clustermap/src/main/java/com.github.ambry.clustermap/HelixBootstrapUpgradeUtil.java

```diff
@@ -94,22 +98,28 @@
   private final String zkLayoutPath;
   private final String stateModelDef;
   private final Map<String, HelixAdmin> adminForDc = new HashMap<>();
-  private final Map<String, Map<DiskId, SortedSet<Replica>>> instanceToDiskReplicasMap = new HashMap<>();
-  private final Map<String, Map<String, DataNodeId>> dcToInstanceNameToDataNodeId = new HashMap<>();
+  // These two maps should be concurrent map because they are accessed in multi-threaded context (see addUpdateInstances
+  // method).
+  // For now, the inner map doesn't need to be concurrent map because it is within a certain dc which means there should
+  // be only one thread that updates it.
+  private final ConcurrentHashMap<String, Map<DiskId, SortedSet<Replica>>> instanceToDiskReplicasMap =
+      new ConcurrentHashMap<>();
+  private final ConcurrentHashMap<String, Map<String, DataNodeId>> dcToInstanceNameToDataNodeId =
+      new ConcurrentHashMap<>();
   private final int maxPartitionsInOneResource;
   private final boolean dryRun;
   private final boolean forceRemove;
-  private int maxResource = -1;
+  private final boolean resourceChangeOnly;
   private final String clusterName;
   private boolean expectMoreInHelixDuringValidate = false;
-  private Map<String, Set<String>> instancesNotForceRemovedByDc = new HashMap<>();
-  private Map<String, Set<String>> partitionsNotForceRemovedByDc = new HashMap<>();
-  private int instancesAdded = 0;
-  private int instancesUpdated = 0;
-  private int instancesDropped = 0;
-  private int resourcesAdded = 0;
-  private int resourcesUpdated = 0;
-  private int resourcesDropped = 0;
+  private ConcurrentHashMap<String, Set<String>> instancesNotForceRemovedByDc = new ConcurrentHashMap<>();
+  private ConcurrentHashMap<String, Set<String>> partitionsNotForceRemovedByDc = new ConcurrentHashMap<>();
+  private final AtomicInteger instancesAdded = new AtomicInteger();
```

Comment:
Since instancesAdded is changed during execution, so does it need to be final?

Reply:
the `final` means the reference doesn't change. The value of this object is allowed to be mutable.